internal/runtime/client/client_test.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/cluster-api/internal/runtime/catalog"
+	fakev1alpha1 "sigs.k8s.io/cluster-api/internal/runtime/catalog/test/v1alpha1"
+	fakev1alpha2 "sigs.k8s.io/cluster-api/internal/runtime/catalog/test/v1alpha2"
+)
+
+func TestClient_httpCall(t *testing.T) {
+	g := NewWithT(t)
+
+	tableTests := []struct {
+		name     string
+		request  runtime.Object
+		response runtime.Object
+		opts     *httpCallOptions
+		wantErr  bool
+	}{
+		{
+			"all nil, err",
+			nil, nil,
+			nil,
+			true,
+		},
+		{
+			"nil catalog, err",
+			&fakev1alpha1.FakeRequest{}, &fakev1alpha1.FakeResponse{},
+			nil,
```

Comment:
```suggestion
			&httpCallOptions{
				catalog: nil,
			},
```
nit, just to make it easier to read